source/s3_meta_request.c

```diff
@@ -310,6 +519,22 @@ PyObject *aws_py_s3_client_make_meta_request(PyObject *self, PyObject *args) {
     meta_request->self_py = py_s3_request;
     Py_INCREF(meta_request->self_py);
 
+    if (recv_filepath) {
+        meta_request->recv_file = fopen(recv_filepath, "wb+");
+    }
+    if (send_filepath) {
+        meta_request->send_file = fopen(send_filepath, "rb+");
+        if (type == AWS_S3_META_REQUEST_TYPE_PUT_OBJECT) {
+            meta_request->input_body = s_input_stream_new_from_file(allocator, send_filepath, meta_request);
+            if (!meta_request->input_body) {
+                PyErr_SetAwsLastError();
+                goto error;
+            }
+            /* rewrite the input stream of the original request */
+            aws_http_message_set_body_stream(http_request, meta_request->input_body);
```

Comment:
oh man, this is so weird... after this set_body_stream() call we have
`HttpRequest._binding.py_body_stream // still pointing to previous python body stream`
but
`HttpRequest._binding.native->body_stream // pointing to updated C body stream `

that's going to lead to something unexpected happening someday

Also, semi-related, we're not doing anything to keep http_request_py alive for the duration of the request. aws_input_stream still doesn't have refcounting. There's weird lifetime stuff with the python HttpRequest keeps the python InputStream alive. So if the python HttpRequest goes out of scope, the underlying aws_http_message will stay alive because s3 has a refcount on it, but the InputStream will get cleaned up which will destroy the non-ref-counted aws_input_stream

ANYWAY, here's my idea:
1) have the metarequest keep the python HttpRequest alive until the operation completes
2) if send_filepath is set, create a NEW aws_http_request that copies the headers/paths/etc of the one being passed in here, but uses a different aws_input_stream
3) if it's not set, you can just keep using the aws_http_request that got passed in

We'll end up copying the header strings, but it seems like a relatively small cost to pay

Reply:
yeah, it's so weird... And it's so hard🤯